cmd/snap-confine/udev-support.c

```diff
@@ -18,301 +18,390 @@
+	 *
+	 * Because udev does not allow for dots in tag names, those are replaced by
+	 * underscores in snapd. We just match that behavior. */
+	char *udev_tag SC_CLEANUP(sc_cleanup_string) = NULL;
+	udev_tag = sc_udev_mangle_security_tag(security_tag);
+
+	/* Use udev APIs to talk to udev-the-daemon to determine the list of
+	 * "devices" with that tag assigned. The list may be empty, in which case
+	 * there's no udev tagging in effect and we must refrain from constructing
+	 * the cgroup as it would interfere with the execution of a program. */
+	struct udev SC_CLEANUP(sc_cleanup_udev) * udev = NULL;
+	udev = udev_new();
+	if (udev == NULL) {
+		die("cannot connect to udev");
 	}
-	// /dev/nvidia-modeset
-	if (stat(nvidia_modeset_path, &sbuf) == 0) {
-		_run_snappy_app_dev_add_majmin(udev_s, nvidia_modeset_path,
-					       major(sbuf.st_rdev),
-					       minor(sbuf.st_rdev));
+	struct udev_enumerate SC_CLEANUP(sc_cleanup_udev_enumerate) * devices =
+	    NULL;
+	devices = udev_enumerate_new(udev);
+	if (devices == NULL) {
+		die("cannot create udev device enumeration");
 	}
-	// /dev/uhid isn't represented in sysfs, so add it to the device cgroup
-	// if it exists and let AppArmor handle the mediation
-	if (stat("/dev/uhid", &sbuf) == 0) {
-		_run_snappy_app_dev_add_majmin(udev_s, "/dev/uhid",
-					       major(sbuf.st_rdev),
-					       minor(sbuf.st_rdev));
+	if (udev_enumerate_add_match_tag(devices, udev_tag) < 0) {
+		die("cannot add tag match to udev device enumeration");
 	}
-	// When CONFIG_TUN=m, /dev/net/tun will exist but using it doesn't
-	// autoload the tun module but also /dev/net/tun isn't udev tagged
-	// until it is loaded. To work around this, if /dev/net/tun exists, add
-	// it unconditionally to the cgroup and rely on AppArmor to mediate the
-	// access. LP: #1859084
-	if (stat("/dev/net/tun", &sbuf) == 0) {
-		_run_snappy_app_dev_add_majmin(udev_s, "/dev/net/tun",
-					       major(sbuf.st_rdev),
-					       minor(sbuf.st_rdev));
+	if (udev_enumerate_scan_devices(devices) < 0) {
+		die("cannot enumerate udev devices");
 	}
-	// add the assigned devices
-	while (udev_s->assigned != NULL) {
-		const char *path = udev_list_entry_get_name(udev_s->assigned);
-		if (path == NULL)
-			die("udev_list_entry_get_name failed");
-		run_snappy_app_dev_add(udev_s, path);
-		udev_s->assigned = udev_list_entry_get_next(udev_s->assigned);
+	/* NOTE: udev_list_entry is bound to life-cycle of the used udev_enumerate */
+	struct udev_list_entry *assigned;
+	assigned = udev_enumerate_get_list_entry(devices);
+	if (assigned == NULL) {
+		/* NOTE: Nothing is assigned, don't create or use the device cgroup. */
+		debug("no devices tagged with %s, skipping device cgroup setup",
+		      udev_tag);
+		return;
 	}
+
+	sc_cgroup_fds SC_CLEANUP(sc_cleanup_cgroup_fds) fds = { -1, -1, -1 };
+	fds = sc_udev_open_cgroup_v1(security_tag);
+	if (fds.cgroup_procs_fd < 0) {
+		debug("cgroup v1 unavailable, skipping device cgroup setup");
+		return;
```

Comment:
This is expected. If you are on a v2 system you are in forced partial confinement.

Reply:
I didn't read the diff right.

I did have a point here but I'd rather iterate on something rather than discuss it without making progress in the code. I will revert to die() and then open a follow-up that explains my point clearly.

This does make sense in an environment where you don't have cgroups at all, but also in that mode you don't have apparmor or other confinement.

Reply:
This is done